Add environment variables to directory substitution

krafczyk opened the pull request:
This will be useful for users wishing to install an external package in their home directory or an external repo in their home directory. They can then specify the location of these items using $home.

citibeth commented:
Any way this could be made to work with `~` and `~user` as well as `$home`?

krafczyk commented:
Well technically it's already just expanding `~`. I'll see whether python can expand `~user` I'm sure it can though.

Actually, is there any reason we can't just expand environment variables? I think it makes sense to keep a set of protected variables like `$spack` and `$tempdir`, but I don't see any reason why it can't also expand other environment variables. We would just have to raise an exception if that variable isn't defined.

adamjstewart commented:
`~user` isn't super useful in my mind. You can just use `/home/user` instead. `$home` is different because it's dynamic. If you have a single `config.yaml` for a site installation, `$home` will be your home directory. As long as `$home` works I'm happy.

citibeth commented:
Good point.
On May 3, 2017 10:44 AM, "Adam J. Stewart" <notifications@github.com> wrote:

> ~user isn't super useful in my mind. You can just use /home/user instead.
> $home is different because it's dynamic. If you have a single config.yaml
> for a site installation, $home will be your home directory. As long as
> $home works I'm happy.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/4084#issuecomment-298932279>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd61WT0eLXfpd4A6316PERRTsl7Nwks5r2JLPgaJpZM4NOX7P>
> .
>

krafczyk commented:
There are sites though which have different locations for different user directories. It wouldn't then be possible to use `$home/../another_user `to refer to another user's home directory. It might be good to have `~user` functionality in those cases.

And again, does anybody object to expanding generic environment variables as well?

hartzell commented:
> You can just use /home/user instead

Not everyone's home dir is in `/home/user` (thumbs up to @krafczyk's response).

Expanding env vars seems useful (thumbs up also).

adamjstewart commented:
> Not everyone's home dir is in `/home/user`

Then you can use `/other-home/user`? The point is that `~user` isn't dynamic whereas `$home` is.

krafczyk commented:
It turns out that ~user works as well as ~. I'm currently trying to find a way to implement this as well as the expansion of general environment variables.

krafczyk commented:
@adamjstewart @hartzell I've rebased this branch, and am interested in getting it merged.

It turns out that os.path.expanduser is already being called on the path. This works to expand both `~` as well as `~user`. It only works at the front of the path though. This shouldn't be a problem however. So I've removed the extra code I wrote to do this expansion.

I've also added code to expand environment variables as well. This should make configuring spack more flexible.

adamjstewart commented:
I'm wondering whether or not it is safe to expand all variables, but I can't think of a reason not to. I liked the first commit as well, so either 1 or 3 is fine.

Is there any documentation that can be updated to reflect this new capability? New features aren't very useful if you and I are the only ones who know about them 😃

krafczyk commented:
Sure, I'll see where I can throw it in the docs.

What do you mean by 1 or 3? Are you saying that the pull as is won't be accepted?

adamjstewart commented:
By 1 or 3 I'm saying the first commit or the third (current) commit. Both are better than the second commit, although the first commit was simpler and more in line with current logic.

I probably won't be the one to merge this PR (I usually handle package PRs) but @alalazo might.

krafczyk commented:
@adamjstewart I decided to go with 3. I've also added some documentation to make it a little clearer how variable substitution works as well as to elucidate these new capabilities.

scheibelp commented on lib/spack/docs/config_yaml.rst:
```diff
@@ -16,25 +16,6 @@ These settings can be overridden in ``etc/spack/config.yaml`` or
 
 .. _config-file-variables:
```
I'm OK with moving this section but ```.. _config-file-variables:``` needs to move with it.

scheibelp commented on lib/spack/docs/configuration.rst:
```diff
@@ -261,3 +261,52 @@ The merged configuration would look like this:
+------------------------------
+
+Spack understands several variables which can be used in config file paths
+where ever they appear. There are three sets of these variables, Special 
```
I think a name should be chosen to indicate these variables are understood by Spack. "special" is too general. My suggestion is "spack-specific variables" but anything more specific is great.

Also nitpick: the order presented here is special, environment, and user variables. But the order of presentation is special, user, environment. I'd prefer the order of the sections matches the order here (in part because user variables differ from environment and special variables).

krafczyk commented:
@scheibelp Sorry for the wait, I've applied the changes you suggested. Let me know if you have more suggestions.

scheibelp commented:
Thanks @krafczyk. FYI I won't get the chance to review these again until Monday

scheibelp commented:
LGTM but can you close and reopen this PR (in the github UI) to restart the tests (one of them failed incorrectly)?

krafczyk commented:
@scheibelp I would, but to me, it says all 6 tests passed!

adamjstewart commented:
If you go on the Travis page (and have write permissions for the repo) you can re-run a single test. I did this, which is why it's passing now.
